### PR TITLE
feat(slider): Convert JS to TypeScript

### PR DESCRIPTION
**Pull request by acdvorak**
Refs #4225

### CODE REVIEW

File: packages/mdc-slider/adapter.ts [hunk 1]
```diff
@@ -21,178 +21,138 @@
...
-  getTabIndex() {}
+  getTabIndex(): number;
 
   /**
    * Registers an event handler on the root element for a given event.
-   * @param {string} type
-   * @param {function(!Event): undefined} handler
    */
-  registerInteractionHandler(type, handler) {}
+  registerInteractionHandler(type: string, handler: EventListener): void;
```
Comment (moog16) [inline packages/mdc-slider/adapter.ts]: all of the `string` and `EventListener` should be `EventType` and `SpecificEventListener`

File: packages/mdc-slider/foundation.ts [hunk 2]
```diff
@@ -21,113 +21,123 @@
...
-import MDCSliderAdapter from './adapter';
-
-import {getCorrectEventName, getCorrectPropertyName} from '@material/animation/index.ts';
+import {getCorrectEventName, getCorrectPropertyName} from '@material/animation/index';
 import {MDCFoundation} from '@material/base/foundation';
+import {SpecificEventListener} from '@material/base/index';
+import {MDCSliderAdapter} from './adapter';
+import {cssClasses, numbers, strings} from './constants';
+
+type UpEventType = 'mouseup' | 'pointerup' | 'touchend';
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: should we also be exporting these types?

File: packages/mdc-slider/foundation.ts [hunk 3]
```diff
@@ -21,113 +21,123 @@
...
+  private hasTrackMarker_ = false;
+  private handlingThumbTargetEvt_ = false;
+  private min_ = 0;
+  private max_ = 100;
+  private step_ = 0;
+  private value_ = 0;
+  private disabled_ = false;
+  private preventFocusState_ = false;
+
+  private readonly thumbContainerPointerHandler_: EventListener;
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: this could also be SpecificEventListener, right?

File: packages/mdc-slider/foundation.ts [hunk 4]
```diff
@@ -137,15 +147,19 @@ class MDCSliderFoundation extends MDCFoundation {
   init() {
     this.isDiscrete_ = this.adapter_.hasClass(cssClasses.IS_DISCRETE);
     this.hasTrackMarker_ = this.adapter_.hasClass(cssClasses.HAS_TRACK_MARKER);
-    DOWN_EVENTS.forEach((evtName) => this.adapter_.registerInteractionHandler(evtName, this.interactionStartHandler_));
-    this.adapter_.registerInteractionHandler('keydown', this.keydownHandler_);
-    this.adapter_.registerInteractionHandler('focus', this.focusHandler_);
-    this.adapter_.registerInteractionHandler('blur', this.blurHandler_);
+
     DOWN_EVENTS.forEach((evtName) => {
+      this.adapter_.registerInteractionHandler(evtName, this.interactionStartHandler_ as EventListener);
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: I think if you change the adapter methods to `SpecificEventListener` you won't need to cast here

File: packages/mdc-slider/foundation.ts [hunk 5]
```diff
@@ -278,10 +281,8 @@ class MDCSliderFoundation extends MDCFoundation {
 
   /**
    * Called when the user starts interacting with the slider
-   * @param {!Event} evt
-   * @private
    */
-  handleDown_(evt) {
+  private handleDown_(downEvent: MouseLikeEvent): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: void unnecessary

File: packages/mdc-slider/foundation.ts [hunk 6]
```diff
@@ -291,76 +292,68 @@ class MDCSliderFoundation extends MDCFoundation {
...
 
+    const moveEventType = MOVE_EVENT_MAP[downEvent.type as DownEventType];
+
     // Note: upHandler is [de]registered on ALL potential pointer-related release event types, since some browsers
     // do not always fire these consistently in pairs.
     // (See https://github.com/material-components/material-components-web/issues/1192)
     const upHandler = () => {
       this.handleUp_();
-      this.adapter_.deregisterBodyInteractionHandler(MOVE_EVENT_MAP[evt.type], moveHandler);
+      this.adapter_.deregisterBodyInteractionHandler(moveEventType, moveHandler as EventListener);
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: these EventListeners don't need to be cast after rename to SpecificEventListener

File: packages/mdc-slider/foundation.ts [hunk 7]
```diff
@@ -291,76 +292,68 @@ class MDCSliderFoundation extends MDCFoundation {
...
+    this.setValueFromEvt_(downEvent);
   }
 
   /**
    * Called when the user moves the slider
-   * @param {!Event} evt
-   * @private
    */
-  handleMove_(evt) {
+  private handleMove_(evt: MouseLikeEvent): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: void not necessary

File: packages/mdc-slider/foundation.ts [hunk 8]
```diff
@@ -291,76 +292,68 @@ class MDCSliderFoundation extends MDCFoundation {
...
+    return (evt as MouseEvent).pageX;
   }
 
   /**
    * Sets the slider value from an event
-   * @param {!Event} evt
-   * @private
    */
-  setValueFromEvt_(evt) {
+  private setValueFromEvt_(evt: MouseLikeEvent): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: void unnecessary

File: packages/mdc-slider/foundation.ts [hunk 9]
```diff
@@ -371,9 +364,8 @@ class MDCSliderFoundation extends MDCFoundation {
 
   /**
    * Handles keydown events
-   * @param {!Event} evt
    */
-  handleKeydown_(evt) {
+  private handleKeydown_(evt: KeyboardEvent): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: void unnecessary

File: packages/mdc-slider/foundation.ts [hunk 10]
```diff
@@ -543,28 +524,24 @@ class MDCSliderFoundation extends MDCFoundation {
 
   /**
    * Toggles the active state of the slider
-   * @param {boolean} active
    */
-  setActive_(active) {
+  private setActive_(active: boolean): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: void unnecessary

File: packages/mdc-slider/foundation.ts [hunk 11]
```diff
@@ -21,113 +21,123 @@
...
+      setMarkerValue: () => undefined,
+      appendTrackMarkers: () => undefined,
+      removeTrackMarkers: () => undefined,
+      setLastTrackMarkersStyleProperty: () => undefined,
+      isRTL: () => false,
+    };
+    // tslint:enable:object-literal-sort-keys
   }
 
+  private rect_: ClientRect | null = null;
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: should this just be `this.rect_!: ClientRect;`? You're using the `!` below, meaning it should never be null.

File: packages/mdc-slider/foundation.ts [hunk 12]
```diff
@@ -543,28 +524,24 @@ class MDCSliderFoundation extends MDCFoundation {
...
     this.active_ = active;
     this.toggleClass_(cssClasses.ACTIVE, this.active_);
   }
 
   /**
    * Toggles the inTransit state of the slider
-   * @param {boolean} inTransit
    */
-  setInTransit_(inTransit) {
+  private setInTransit_(inTransit: boolean): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: remove void

File: packages/mdc-slider/foundation.ts [hunk 13]
```diff
@@ -543,28 +524,24 @@ class MDCSliderFoundation extends MDCFoundation {
...
     this.toggleClass_(cssClasses.IN_TRANSIT, this.inTransit_);
   }
 
   /**
    * Conditionally adds or removes a class based on shouldBePresent
-   * @param {string} className
-   * @param {boolean} shouldBePresent
    */
-  toggleClass_(className, shouldBePresent) {
+  private toggleClass_(className: string, shouldBePresent: boolean): void {
```
Comment (moog16) [inline packages/mdc-slider/foundation.ts]: remove void

File: packages/mdc-slider/index.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  stepUp(amount = (this.step || 1)) {
+    this.value += amount;
+  }
+
+  stepDown(amount = (this.step || 1)) {
+    this.value -= amount;
+  }
+
+  private parseFloat_(str: string | null, defaultValue: number) {
+    const num = parseFloat(str!); // tslint:disable-line:ban
```
Comment (moog16) [inline packages/mdc-slider/index.ts]: i removed this `tslint:disable-line` - i dont think you need it

File: packages/mdc-slider/adapter.ts (hunk 1 above)
Comment (acdvorak) [reply packages/mdc-slider/adapter.ts]: Done

File: packages/mdc-slider/foundation.ts (hunk 3 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Done

File: packages/mdc-slider/foundation.ts (hunk 4 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Done

File: packages/mdc-slider/foundation.ts (hunk 5 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Removed

File: packages/mdc-slider/foundation.ts (hunk 6 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Done

File: packages/mdc-slider/foundation.ts (hunk 7 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Removed

File: packages/mdc-slider/foundation.ts (hunk 9 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Removed

File: packages/mdc-slider/foundation.ts (hunk 11 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: Done

File: packages/mdc-slider/index.ts (hunk 14 above)
Comment (acdvorak) [reply packages/mdc-slider/index.ts]: When I remove it, I get:

```
ERROR: packages/mdc-slider/index.ts:178:17 - Calls to 'parseFloat' are not allowed. tsstyle#type-coercion
```

File: packages/mdc-slider/foundation.ts (hunk 2 above)
Comment (acdvorak) [reply packages/mdc-slider/foundation.ts]: They're only used internally within the foundation; none of our public APIs use them. So I think we're probably safe keeping them private.

File: packages/mdc-slider/foundation.ts [hunk 15]
```diff
@@ -291,76 +287,68 @@ class MDCSliderFoundation extends MDCFoundation {
...
    * Computes the new value from the pageX position
-   * @param {number} pageX
-   * @return {number}
    */
-  computeValueFromPageX_(pageX) {
+  private computeValueFromPageX_(pageX: number): number {
     const {max_: max, min_: min} = this;
-    const xPos = pageX - this.rect_.left;
-    let pctComplete = xPos / this.rect_.width;
+    const xPos = pageX - this.rect_!.left; // rect_ is assigned in layout()
```
Comment (kfranqueiro) [inline packages/mdc-slider/foundation.ts]: Are the `!`s really needed on `rect_` everywhere even though the member declaration already has it?

Comment (moog16) [reply packages/mdc-slider/foundation.ts]: ah good catch. Those should be removed too.

**Pull request merged**